source/agora/node/Ledger.d

```diff
@@ -395,6 +395,45 @@ public class Ledger
 
         // remove the TXs from the Pool
         block.txs.each!(tx => this.pool.remove(tx));
+
+        this.updateSlashedUTXOSet(block);
+    }
+
+    /***************************************************************************
+
+        Update the UTXOs of validators that are to be slashed
+
+        Params:
+            block = the block to update the UTXO set with
+
+    ***************************************************************************/
+
+    protected void updateSlashedUTXOSet (const ref Block block) @safe
+    {
+        Hash[] validator_utxos;
+        this.slash_man.getMissingValidatorsUTXOs(validator_utxos,
+            block.header.missing_validators);
+        foreach (utxo; validator_utxos)
+        {
+            UTXO utxo_value;
+            if (!this.utxo_set.peekUTXO(utxo, utxo_value))
+                assert(0, "UTXO for the slashed validator not found!");
+
+            auto remain_amount = Amount(utxo_value.output.value);
+            remain_amount.sub(this.slash_man.penalty_amount);
+            Transaction slashing_tx =
+            {
+                TxType.Payment,
+                inputs: [Input(utxo)],
+                outputs: [
+                    Output(this.slash_man.penalty_amount,
+                        this.slash_man.penalty_address),
+                    Output(remain_amount, utxo_value.output.address),
```

Comment:
This means if the node gets punished then parts of their funds will be sent to the commons budget, but now the leftover part will be immediately unfrozen. I'm not sure this is the best idea. I would have hoped that the leftover part either remains frozen, or becomes a melting transaction.

But I understand it might be complicated to implement.

How about we leave it as is for now and we file a new issue for improving this later?

Reply:
> How about we leave it as is for now and we file a new issue for improving this later?

Yes, it could be implemented with a more reasonable policy about the frozen UTXO.